### DISCUSSION TITLE
Behind the Scene: Understanding the process of running test and the verification after git push

### DISCUSSION THREAD

**Discussion by NimaSarajpoor** [TDAmeritrade/stumpy#390] [Q&A]
Hello everyone,

I am trying to understand what's going on behind the scene and I would truly appreciate if someone could help me out with the following questions:

* I don't know why the TDAmeritrade.stumpy is not going to be successful here in this repo whenever I do git push. I checked one of the errors before and it is apparently related to the running time If my understanding was correct. Any idea how to fix that? (should I be worried regarding that matter?) 

* I also noticed that it skipped one of the test and it says: no GPU available. However, I think I have one! 
![image](https://user-images.githubusercontent.com/38519522/119894289-a9cf4780-bef9-11eb-81f6-2ca63a39b68f.png)
__So, should I somehow activate GPU since it cannot be recognized by stumpy?__ If I remember correctly, I did it once a few months ago on another project where I had to install cudatoolkit.

* The unit test is fine on my end. 
__NOTE:__ I, however, should confess that for the last commit that I pushed (ffde197), I forgot to do the unit test first. I did it after the git push. According to the messages, the tests were successful on my local repo.
But, I got error when the unit test is performed on stumpy repo after I did git push. (please recall that I didn't do the `./setup.sh && ./test.sh` on my local repo before the `git push ...`.  I checked the error provided on the stumpy repo and I found this:
![image](https://user-images.githubusercontent.com/38519522/119887451-696bcb80-bef1-11eb-8c17-29655571a70b.png)

Is this because my branch was probably based on an old-version of the master(origin) when I did git push? So, when I forgot to do the setup and unit test prior to the git push, the unit test performed on this repo was actually testing the previous version of the files and NOT the most-updated ones?

* After running `flake8` and `black` and `./setup.sh && ./test.sh`  in my shell, I should say: 
** running `flake8` returns nothing
** running `black` returns the following message:  __No Path provided. Nothing to do \U0001f634__
** running `./setup.sh && ./test.sh` shows that the scripts run both `flake8` and `black` before doing the unit test. So, what is the point of running `flake8` and `black`  separately in the first place? __Isn't it redundant?__

**Comment by seanlaw** [comment]
> I don't know why the TDAmeritrade.stumpy is not going to be successful here in this repo whenever I do git push. I checked one of the errors before and it is apparently related to the running time If my understanding was correct. Any idea how to fix that? (should I be worried regarding that matter?)

I'm not sure what you mean here. There is not enough information to provide a response. Can you please clarify?  It is often best to provide the single command and it's full Python traceback/error

> I also noticed that it skipped one of the test and it says: no GPU available. However, I think I have one! So, should I somehow activate GPU since it cannot be recognized by stumpy? If I remember correctly, I did it once a few months ago on another project where I had to install cudatoolkit.

Yes, you need to make sure that the cudatoolkit is installed properly and compatible with your system in order for STUMPY to be able to pick it up. `./conda.sh` attempts to install the `cudatoolkit` for you but there may be some incompatibilities with Windows machines and, unfortunately, setting this up is beyond the scope of STUMPY.

> The unit test is fine on my end.
NOTE: I, however, should confess that for the last commit that I pushed (ffde197), I forgot to do the unit test first. I did it after the git push. According to the messages, the tests were successful on my local repo.
But, I got error when the unit test is performed on stumpy repo after I did git push. (please recall that I didn't do the ./setup.sh && ./test.sh on my local repo before the git push .... I checked the error provided on the stumpy repo and I found this:

Where is that error message from? Sometimes, the Azure Pipelines tests have issues so please let me worry about them (sometimes, the tests just need to be restarted). Just make sure that your local testing passes without error before pushing. 

> After running flake8 and black and ./setup.sh && ./test.sh in my shell, I should say:
> ** running ./setup.sh && ./test.sh shows that the scripts run both flake8 and black before doing the unit test. So, what is the point of running flake8 and black separately in the first place? Isn't it redundant?

So, running `flake8 ./` beforehand is redundant but running `black ./` beforehand is NOT redundant. `test.sh` does indeed run `flake8 ./` but it runs `black --check --diff ./`, which only checks the formatting but does NOT change any of the Python code files. Running `black ./` will actually change/fix any poorly formatted Python code (i.e., only files that end in `.py`) while `black --check --diff ./` will only check the formatting and errors out if there is poor code formatting. Neither `flake8 ./` or `black ./` will do anything with `.ipynb` files. To some extent, since you are currently only changing/adding `.ipynb` files, running the unit tests actually doesn't make any sense (it's good practice though). When you start adding/changing `.py` files (including adding/changing unit tests) then testing will really be important.

Let me know if that makes sense and great questions!

**Reply by NimaSarajpoor** [reply]
Thanks for your response. I should have provided the message. Thanks for explaining those things.
